src/nifreeze/estimator.py

```diff
@@ -22,229 +22,151 @@
 #
 """A model-based algorithm for the realignment of dMRI data."""
 
+from __future__ import annotations
+
 from pathlib import Path
-from tempfile import TemporaryDirectory, mkstemp
+from tempfile import TemporaryDirectory
+from typing import Self
 
-import nibabel as nb
 from tqdm import tqdm
 
-from nifreeze.data.splitting import lovo_split
-from nifreeze.model.base import ModelFactory
-from nifreeze.registration.ants import _prepare_registration_data, _run_registration
+from nifreeze.data.base import BaseDataset
+from nifreeze.model.base import BaseModel, ModelFactory
+from nifreeze.registration.ants import (
+    _prepare_registration_data,
+    _run_registration,
+)
 from nifreeze.utils import iterators
 
 
+class Filter:
+    """Alters an input data object (e.g., downsampling)."""
+
+    def run(self, dataset: BaseDataset, **kwargs):
+        """
+        Trigger execution of the designated filter.
+
+        Parameters
+        ----------
+        dataset : :obj:`~nifreeze.data.base.BaseDataset`
+            The input dataset this estimator operates on.
+
+        Returns
+        -------
+        :obj:`~nifreeze.estimator.Estimator`
+            The estimator, after fitting.
+
+        """
+        return dataset
+
+
 class Estimator:
     """Estimates rigid-body head-motion and distortions derived from eddy-currents."""
 
-    @staticmethod
-    def estimate(
-        data,
-        *,
-        align_kwargs=None,
-        iter_kwargs=None,
-        models=("b0",),
-        omp_nthreads=None,
-        n_jobs=None,
+    __slots__ = ("_model", "_strategy", "_prev", "_model_kwargs", "_align_kwargs")
+
+    def __init__(
+        self,
+        model: BaseModel | str,
+        strategy: str = "random",
+        prev: Self | None = None,
```

Comment:
I think this means that `EstimatorA(prev=EstimatorA())` works but `EstimatorB(prev=EstimatorB())` won't. If any subclass should work, then:

```suggestion
        prev: Estimator | Filter | None = None,
```

Also adding `Filter`, since that seems a design goal.